### ISSUE TITLE
Using hnsw and lsh as classifier

### ISSUE THREAD

**Issue by QueSabz** [VarIr/scikit-hubness#86]
How can I use both methods for data classification rather than just retrieving nearest neighbours.
Or should I just use it like this kNN = KNeighborsClassifier(n_neighbors=50, algorithm = m) where m can be hnsw or lsh?

**Comment by VarIr** [comment]
I assume you want to compare the two methods with respect to your classification task?
In this case, I would approach this exactly the way you suggested. Simply create two `KNeighborsClassifier`s, 
one with `algorithm="hnsw"` and one with `algorithm="lsh"`. 

Note that `v0.30` is currently under development. It's not yet finished, but it should be possible to install it from the main branch here. It integrates nicely with recent versions of scikit-learn that provide the [`KNeighborsTransformer`](https://scikit-learn.org/stable/modules/generated/sklearn.neighbors.KNeighborsTransformer.html#sklearn.neighbors.KNeighborsTransformer). That is, you can precompute k-neighbors graphs with various (approximate) neighbor methods with scikit-hubness, and feed the results to any scikit-learn function that works on such graphs, including classification.

**Comment by QueSabz** [comment]
I have installed it and it seems to be working but when I import it on my notebooj I get the following warnings:
![2022-03-14 12_21_52-SpeedUp Methods - Jupyter Notebook](https://user-images.githubusercontent.com/48109511/158171422-baf2636d-4164-4259-b4b0-9d13865ef43c.png)

**Comment by VarIr** [comment]
These are known issues of an old scikit-learn version when used with newer numpy versions.
They do not cause any errors and you may ignore them, or even filter them away:

``` python
import warnings

with warnings.catch_warnings():
    warnings.filterwarnings("ignore", category=DeprecationWarning)
    from sklearn import ...  # Insert all your sklearn-related imports here
```

These warnings will also go away with scikit-hubness v0.30 which works with scikit-learn versions > 0.21.

**Comment by QueSabz** [comment]
How do I upgrade to scikit-hubness v0.30? I tried using the following command on anaconda :
 pip install -U scikit-hubness but I got the followin
![2022-03-14 13_03_32-C__Windows_system32_cmd exe](https://user-images.githubusercontent.com/48109511/158177503-5db04950-5fa3-497d-b07b-f127643ea00e.png)
g

**Comment by VarIr** [comment]
A note of caution first, v0.30 is experimental at this point and might have all kinds of problems as is.
If you still want to try it, I suggest creating a new conda environment and installing it there.

``` bash
# Install setup requirements
python3 -m pip install --upgrade pip
python3 -m pip install setuptools wheel pybind11

# Install scikit-hubness from main branch; including approx. neighbors libraries
python3 -m pip install git+https://github.com/VarIr/scikit-hubness.git[ann]
```

**Comment by QueSabz** [comment]
![2022-03-14 13_46_59-C__Windows_system32_cmd exe](https://user-images.githubusercontent.com/48109511/158185170-4e203081-a6e8-4854-9cb6-7cc1d85a5c04.png)

**Comment by VarIr** [comment]
Can you please try without the `[ann]` part in the end? Perhaps this does not work when installing directly from github. The difference would only be whether other packages like `nmslib` or `annoy` are installed automatically at the same time. To use HNSW, you would then also have to `pip install nmslib`.

**Comment by QueSabz** [comment]
I have done all the steps in a new environment as suggested by when importing from skhubness I get the following error message:
![image](https://user-images.githubusercontent.com/48109511/158418928-345545c8-dd90-4f42-8905-7e9692deb7e0.png)

**Comment by VarIr** [comment]
From v0.30 scikit-hubness does provide an own implementation of KNeighborsClassifier anymore.
Instead, you can use the one from scikit-learn. The workflow will be as outline [here](https://scikit-learn.org/stable/modules/neighbors.html#nearest-neighbors-transformer). For example, you can create a pipeline of NMSlibTransformer and KNeighborsClassifier.

**Comment by QueSabz** [comment]
In terms of pipeline creation I am not well versed. Is there any possibility I can see how this is done for lsh and HNSW classification methods as I really want to use them to do comparison in terms of time complexity and accuracy. For example I have already used the exact methods which are brute force, kd tree, and ball tree in this manner:
![image](https://user-images.githubusercontent.com/48109511/158428821-8c1a5624-4eb7-4f21-bf52-ecd7854175c7.png)

**Comment by VarIr** [comment]
I would in this case suggest you stick with the old scikit-hubness version for the time being,
because I don't really have any tested example code for the new version at this point.
It will take a while until I manage to update documentation and examples.

**Comment by QueSabz** [comment]
Okay thanks. So with the old version there is no problem to use both lsh and hnsw as classifiers?
I can have some thing like this: models = ['brute','kd_tree', 'ball_tree', 'lsh', 'hnsw']

**Comment by QueSabz** [comment]
![2022-03-17 16_17_49-Approximate Nearest Neighbour - Jupyter Notebook](https://user-images.githubusercontent.com/48109511/158845948-0dd76c70-efe6-42b4-a74c-42512fb5f8d9.png)

Is there a way of installing and importing puffinn package because I want to use it for lsh method
I have installed it using: pip install puffin
But still when I run my notebook I get error above

**Comment by VarIr** [comment]
`puffinn` (note the double-n) is not available from PyPI. `puffin` is an unrelated package.
Did you get any warnings/error messages from following the manual puffinn installation steps?
If no, did you install it in the same conda environment you are using in this Jupyter notebook?

**Comment by QueSabz** [comment]
![image](https://user-images.githubusercontent.com/48109511/159255537-366b54e4-44d2-4bb3-a5a2-d99e04954b34.png)
I followed instruction from this page and hnsw is working perfectly with nmslib installed but its a different story for lsh

**Comment by VarIr** [comment]
I only now realized you are on Windows. Unfortunately, `falconn` does not support Windows, and neither does `puffinn`. So LSH is only available on Linux or Macos.

**Comment by QueSabz** [comment]
Do you have any idea How can I modify/add scoring method on the following annoy wrapper class so that it can fully function as a classifier:
                    ```#https://pypi.org/project/annoy/
#Based on code from: https://scikit-learn.org/stable/auto_examples/neighbors/approximate_nearest_neighbors.html
import annoy 
from collections import Counter
from sklearn.base import BaseEstimator
#Wrapper for using annoy.AnnoyIndex as sklearn's KNeighborsTransformer
class Annoy():
    
    def __init__(self,n_neighbors=5, metric='euclidean', n_trees=10):
        self.n_neighbors = n_neighbors
        self.Ametric = metric
        self.n_trees = n_trees
    
    def fit(self, X_train, y_train):
        self.N_feat = X_train.shape[1]
        self.N_train = X_train.shape[0]
        self.y_train = y_train
        self.t = annoy.AnnoyIndex(self.N_feat,metric=self.Ametric)
        for i, v in zip(range(self.N_train), X_train):
            self.t.add_item(i, v)
        self.t.build(self.n_trees)
        return self

    def predict(self,X_test):
        y_pred = []
        for tv in X_test:
            nn_inds = self.t.get_nns_by_vector(tv, self.n_neighbors)
            nn_classes =[self.y_train[nn] for nn in nn_inds]
            y_hat.append(most_frequent(nn_classes))
        return y_pred
    
def most_frequent(List): 
    occurence_count = Counter(List) 
    return occurence_count.most_common(1)[0][0] 
```

**Comment by VarIr** [comment]
You could try inheriting from the [`ClassifierMixin`](https://github.com/scikit-learn/scikit-learn/blob/e113897235feaf309eaaed24001ca96f3608602f/sklearn/base.py#L635) which provides the `score()` method. Or explicitly write the method. If you take a look at the linked code, it's literally two lines of code.

**Comment by QueSabz** [comment]
Right no I am using UBUNTU but still when I am installing the sk-hubness package it breaks when building and installing falconn package. Are there any specific steps/commands I need to follow when using ubuntu?

**Comment by VarIr** [comment]
Please post the error message you get.

**Comment by QueSabz** [comment]
![Screenshot from 2022-04-01 13-17-26](https://user-images.githubusercontent.com/48109511/161253654-4f57d671-2eea-4e7d-88e8-b399180cc677.png)

**Comment by QueSabz** [comment]
![Screenshot from 2022-04-01 13-21-09](https://user-images.githubusercontent.com/48109511/161253871-d1e815a2-235e-4aac-8ef8-1a44798e4ffc.png)

**Comment by QueSabz** [comment]
![Screenshot from 2022-04-01 13-23-51](https://user-images.githubusercontent.com/48109511/161254253-de1fc9a0-d34a-48ac-9773-4b6b36fc7820.png)

**Comment by VarIr** [comment]
I quickly googled these error messages and some reports on insufficient memory popped up. But this is a wild guess.
I fear I cannot give more support on falconn compilation bugs. You may try to reach out to the [falconn](https://github.com/FALCONN-LIB/FALCONN) devs about this. (Note, however, that falconn has seen no development for over 4 years, for which reason I am dropping falconn support completely with v0.30).

**Comment by VarIr** [comment]
With last activity in this thread nearly one year ago, I'll close this now. Feel free to open a new issue, if further questions arise.
